extensions/amp-auto-ads/0.1/placement.js

```diff
@@ -0,0 +1,202 @@
+};
+
+export class Placement {
+  /**
+   * @param {!Window} win
+   * @param {!Element} anchorElement
+   * @param {!function(!Element, !Element)} injector
+   */
+  constructor(win, anchorElement, injector) {
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    /** @const @private {!Element} */
+    this.anchorElement_ = anchorElement;
+
+    /** @const @private {!function(!Element, !Element)} */
+    this.injector_ = injector;
+
+    /** @private {?Element} */
+    this.adElement_ = null;
+
+    /** @private {!PlacementState} */
+    this.state_ = PlacementState.UNUSED;
+  }
+
+  /**
+   * @param {string} type
+   * @param {!Array<!DataAttributeDef>} dataAttributes
+   * @return {!Promise<!PlacementState>}
+   */
+  placeAd(type, dataAttributes) {
+    this.adElement_ = this.createAdElement_(type, dataAttributes);
+    this.injector_(this.anchorElement_, this.adElement_);
+    return resourcesForDoc(this.adElement_).attemptChangeSize(
+        this.adElement_, TARGET_AD_HEIGHT_PX, undefined).then(() => {
+          this.state_ = PlacementState.PLACED;
+          return this.state_;
+        }).catch(() => {
+          this.state_ = PlacementState.RESIZE_FAILED;
+          return this.state_;
+        });
+  }
+
+  /**
+   * @param {string} type
+   * @param {!Array<!DataAttributeDef>} dataAttributes
+   * @return {!Element}
+   * @private
+   */
+  createAdElement_(type, dataAttributes) {
+    const attributes = {
+      type,
+      'layout': 'responsive',
+      'width': '320',
+      'height': '0',
+    };
+    for (let i = 0; i < dataAttributes.length; ++i) {
+      attributes['data-' + dataAttributes[i].name] = dataAttributes[i].value;
+    }
+    return createElementWithAttributes(
+        this.win_.document, 'amp-ad', attributes);
+  }
+}
+
+/**
+ * @param {!Window} win
+ * @param {!JSONType} configObj
+ * @return {!Array<!Placement>}
+ */
+export function getPlacementsFromConfigObj(win, configObj) {
+  const placementObjs = configObj['placements'];
+  if (!placementObjs) {
+    dev().warn(TAG, 'No placements in config');
+    return [];
+  }
+  const placements = [];
+  for (let i = 0; i < placementObjs.length; ++i) {
+    const placement = getPlacementFromObject(win, placementObjs[i]);
+    if (placement) {
+      placements.push(placement);
+    }
+  }
+  return placements;
+}
+
+/**
+ * Validates that the placementObj represents a valid placement and if so
+ * constructs and returns an instance of the Placement class for it.
+ * @param {!Window} win
+ * @param {!Object} placementObj
+ * @return {?Placement}
+ */
+function getPlacementFromObject(win, placementObj) {
+  const anchor = placementObj['anchor'];
+  if (!anchor) {
+    dev().warn(TAG, 'No anchor in placement');
+    return null;
+  }
+  const anchorElement = getAnchorElement(win, anchor);
+  if (!anchorElement) {
+    dev().warn(TAG, 'No anchor element found');
+    return null;
+  }
+  const injector = INJECTORS[placementObj['pos']];
```

Comment:
We can move this before the anchor element stuff, and save DOM operations if this is null.